src/libfuncs/bounded_int.rs

```
@@ -0,0 +1,739 @@
+        }
+    }
+}
+
+/// Generate MLIR operations for the `bounded_int_add` libfunc.
+#[allow(clippy::too_many_arguments)]
+fn build_add<'ctx, 'this>(
+    context: &'ctx Context,
+    registry: &ProgramRegistry<CoreType, CoreLibfunc>,
+    entry: &'this Block<'ctx>,
+    location: Location<'ctx>,
+    helper: &LibfuncHelper<'ctx, 'this>,
+    _metadata: &mut MetadataStorage,
+    info: &SignatureOnlyConcreteLibfunc,
+) -> Result<()> {
+    let lhs_value = entry.argument(0)?.into();
+    let rhs_value = entry.argument(1)?.into();
+
+    // Extract the ranges for the operands and the result type.
+    let lhs_ty = registry.get_type(&info.signature.param_signatures[0].ty)?;
+    let rhs_ty = registry.get_type(&info.signature.param_signatures[1].ty)?;
+
+    let lhs_range = lhs_ty.integer_range(registry).unwrap();
+    let rhs_range = rhs_ty.integer_range(registry).unwrap();
+    let dst_range = registry
+        .get_type(&info.signature.branch_signatures[0].vars[0].ty)?
+        .integer_range(registry)
+        .unwrap();
+
+    let lhs_width = if lhs_ty.is_bounded_int(registry) {
+        lhs_range.offset_bit_width()
+    } else {
+        lhs_range.zero_based_bit_width()
+    };
+    let rhs_width = if rhs_ty.is_bounded_int(registry) {
+        rhs_range.offset_bit_width()
+    } else {
+        rhs_range.zero_based_bit_width()
+    };
+
+    // Calculate the computation range.
+    let compute_range = Range {
+        lower: (&lhs_range.lower)
+            .min(&rhs_range.lower)
+            .min(&dst_range.lower)
+            .clone(),
+        upper: (&lhs_range.upper)
+            .max(&rhs_range.upper)
+            .max(&dst_range.upper)
+            .clone(),
+    };
+    let compute_ty = IntegerType::new(context, compute_range.offset_bit_width()).into();
+
+    // Zero-extend operands into the computation range.
+    assert!(compute_range.offset_bit_width() >= lhs_width);
```

Can we return an error here?